### PR TITLE
Task 8

### PR DESCRIPTION
**Pull request by DavidCobb606**
Task-8 finished

### CODE REVIEW

File: __tests__/api.test.js [hunk 1]
```diff
@@ -151,5 +149,85 @@ describe("PATCH /api/articles/:article_id",() => {
...
+            comment_count: expect.any(String)
+        })
+      )})    
+    })      
+    })
+    test("The array of objects should be sorted by the date of created, going from most recent to least recent", () => {
+      return request(app).get("/api/articles")
+      .then(({body}) => {
+        const articles = body.articles
+        expect(articles).toBeSortedBy("created_at", {descending: true})
```
Comment (itskatherine) [inline __tests__/api.test.js:182]: I'm wondering how jest sorted is working if there's no require("jest-sorted") at the start...
Could you please add it in and check this still passes?

File: __tests__/api.test.js [hunk 2]
```diff
@@ -151,5 +149,85 @@ describe("PATCH /api/articles/:article_id",() => {
...
+            title: expect.any(String),
+            comment_count: expect.any(String)
+        })
+      )})    
+    })      
+    })
+    test("The array of objects should be sorted by the date of created, going from most recent to least recent", () => {
+      return request(app).get("/api/articles")
+      .then(({body}) => {
+        const articles = body.articles
```
Comment (itskatherine) [inline __tests__/api.test.js:181]: Could use destructuring here, but fine as is!
Equivalent destructured would be
const {articles} = body;

File: __tests__/api.test.js [hunk 3]
```diff
@@ -151,5 +149,85 @@ describe("PATCH /api/articles/:article_id",() => {
...
+        expect(articles).toBeSortedBy("created_at", {descending: true})
+      })
+    })
+    test("The endpoint /api/articles should accept the query `topic`, which filters the article by whichever topic the client chooses in the query.", () => {
+      return request(app)
+      .get("/api/articles/?topic=cats")
+      .then(({body}) => {
+
```
Comment (itskatherine) [inline __tests__/api.test.js:189]: Do need to check the length of this array again, to check it's not empty. (So if there was 3 articles about cats, we would expect the length to be 3)

File: controllers.js [hunk 4]
```diff
@@ -42,6 +42,30 @@ exports.getModifiedArticle = (req,res,next) =>{
...
+
+exports.getArticles =  (req,res,next) => {
+   
+    const {topic} = req.query
+    
+  const lengthCheck = Object.keys(req.query).length
+ 
+
+  if(lengthCheck === 0 || (lengthCheck === 1 && req.query.topic !== undefined)){
+
```
Comment (itskatherine) [inline controllers.js:54]: Any checks of the data are best done in the model, the controller's role is just to extract the data and pass it to the model.
If what it passes is undefined, that can be dealt with in the model.

File: __tests__/api.test.js [hunk 5]
```diff
@@ -151,5 +149,85 @@ describe("PATCH /api/articles/:article_id",() => {
...
+      .get("/api/articles/?topic=computers")
+      .expect(404)
+      .then(({body}) => {
+        expect(body.msg).toBe("Not Found")
+      })
+
+    })
+
+  })
+
```
Comment (itskatherine) [inline __tests__/api.test.js:229]: We need an additional test to check that if we enter a valid topic, but it has no articles, it will return an empty array.
In order for this to work, you will need to check in your model if the topic exists.

**Review by itskatherine — COMMENTED** [review]
Hey David, great start on this ticket with a few additions needed - remember to check for the lengths of the array when you use a forEach in a test 👍 
Could you please have a look at these changes and once you've done them put in another pull request for this ticket
